util/src/testing.rs

```diff
@@ -0,0 +1,95 @@
+use core::fmt;
+use core::mem;
+use core::slice;
+
+/// Internal definition type used for a test.
+pub struct Test {
+    pub module: &'static str,
+    pub name: &'static str,
+    pub run: fn() -> bool,
+}
+
+/// Type which may be used as a test return type.
+pub trait TestResult {
+    /// Report any errors to `tracing`, then returns either `true` for a
+    /// success, or `false` for a failure.
+    fn report(self) -> bool;
+}
+
+impl TestResult for () {
+    fn report(self) -> bool {
+        true
+    }
+}
+
+impl<T: fmt::Debug> TestResult for Result<(), T> {
+    fn report(self) -> bool {
+        match self {
+            Ok(_) => true,
+            Err(err) => {
+                tracing::error!("FAIL {:?}", err);
+                false
+            }
+        }
+    }
+}
+
+/// Declare a new test, sort-of like the `#[test]` attribute.
+// FIXME: Declare a `#[test]` custom attribute instead?
+#[macro_export]
+macro_rules! decl_test {
+    (fn $name:ident $($t:tt)*) => {
+        fn $name $($t)*
+
+        // Introduce an anonymous const to avoid name conflicts. The `#[used]`
+        // will prevent the symbol from being dropped, and `link_section` will
+        // make it visible.
+        const _: () = {
```

Comment:
Unfortunately, we can't put this behind a `#[cfg(test)]`, as that cfg flag won't be visible in dependencies. We may need to use `RUSTFLAGS="--cfg mycelium_test"` (or something like that) to turn on test code for the whole dependency tree.

Reply:
that makes sense, i would really like to avoid linking these into non-test builds eventually. this is fine for now...